### PR TITLE
fix: add test with skipper, kubernetes and redis autoscaling updates

### PR DESCRIPTION
**Pull request by szuecs**
test: add test with skipper, kubernetes and redis autoscaling updates via kubernetes
test: enable changing redis update interval for testing purposes refactor: expose Endpoint and EndpointList
refactor: vegetaAttacker into httptest client
feature: redis can listen on any port add test conditions

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

### CODE REVIEW

**Comment by szuecs** [comment]
waits for https://github.com/zalando/skipper/pull/2080

File: dataclients/kubernetes/clusterstate.go [hunk 1]
```diff
@@ -55,6 +51,26 @@ func (state *clusterState) GetEndpointsByService(namespace, name, protocol strin
...
+	targets := ep.targetsByServicePort(protocol, servicePort)
+	sort.Strings(targets)
+	state.cachedEndpoints[epID] = targets
+	return targets
+}
+
+func (state *clusterState) GetEndpointsByName(namespace, name, protocol string) []string {
+	epID := endpointID{
+		ResourceID: newResourceID(namespace, name),
+		Protocol:   protocol,
```
Comment (RomanZavodskikh) [inline dataclients/kubernetes/clusterstate.go:72]: Why can we avoid initializing Port here?

Comment (RomanZavodskikh) [inline dataclients/kubernetes/clusterstate.go:72]: Why can we just avoid initializing Port here?

File: dataclients/kubernetes/ingressdefinitions.go [hunk 2]
```diff
@@ -138,6 +139,20 @@ func (ep endpoint) targetsByServiceTarget(protocol string, serviceTarget *defini
 	return nil
 }
 
+func (ep Endpoint) targets(protocol string) []string {
+	result := make([]string, 0)
+	for _, s := range ep.Subsets {
+		for _, p := range s.Ports {
+			for _, a := range s.Addresses {
+				result = append(result, formatEndpoint(a, p, protocol))
```
Comment (RomanZavodskikh) [inline dataclients/kubernetes/ingressdefinitions.go:147]: So, would we append to result every possible addr-port pair even if there is no opened port with corresponding number on corresponding address?

File: dataclients/kubernetes/kube.go [hunk 3]
```diff
@@ -363,11 +364,14 @@ func mapRoutes(r []*eskip.Route) map[string]*eskip.Route {
 }
 
 func (c *Client) loadAndConvert() ([]*eskip.Route, error) {
+	c.mu.Lock()
```
Comment (RomanZavodskikh) [inline dataclients/kubernetes/kube.go:367]: Why not defer?

Comment (szuecs) [reply dataclients/kubernetes/kube.go:367]: Because the function does too much and I want to reduce holding the lock

File: dataclients/kubernetes/ingressdefinitions.go (hunk 2 above)
Comment (szuecs) [reply dataclients/kubernetes/ingressdefinitions.go:147]: yes if we would have 10 redis pods with 2 ports we would have 20 endpoints, on the other hand I would expect that redis is only having one port in the service/endpoint and for testing this is more handy, because with testcontainers you get a random port and always localhost as ip, so we need to be able to have all the ports for testing purposes.

File: dataclients/kubernetes/clusterstate.go (hunk 1 above)
Comment (szuecs) [reply dataclients/kubernetes/clusterstate.go:72]: I don't see port here

Comment (szuecs) [reply dataclients/kubernetes/clusterstate.go:72]: I don't see port here

Comment (RomanZavodskikh) [reply dataclients/kubernetes/clusterstate.go:72]:
```
type endpointID struct {
	definitions.ResourceID
	TargetPort string
	Protocol   string
}
```
I see TargetPort here. However, I don't see it initialized here.

Comment (szuecs) [reply dataclients/kubernetes/clusterstate.go:72]: This func will do Endpoint lookup by name and namespace which identifies the kubernetes object.
TargetPort makes no sense in case we want all ports and not one specific port. For tests we use random ports on 127.0.0.1 so we can only listen on multiple ports, so targetport makes no sense in this case.

File: dataclients/kubernetes/ingressdefinitions.go [hunk 4]
```diff
@@ -74,13 +75,13 @@ func (s service) getTargetPortByValue(p int) (*definitions.BackendPort, bool) {
 	return nil, false
 }
 
-type endpoint struct {
+type Endpoint struct {
```
[the remark this answers is not in the document]
Comment (RomanZavodskikh) [reply dataclients/kubernetes/ingressdefinitions.go:78]: Discussed with @szuecs , he told me he can roll back this change.

[the remark this answers is not in the document]
Comment (szuecs) [reply dataclients/kubernetes/ingressdefinitions.go]: fixed

**Comment by RomanZavodskikh** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

File: dataclients/kubernetes/clusterstate.go [hunk 5]
```diff
@@ -22,8 +22,6 @@ type clusterState struct {
 }
 
 func (state *clusterState) getService(namespace, name string) (*service, error) {
-	state.mu.Lock()
-	defer state.mu.Unlock()
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/clusterstate.go:26]: Why is this removed if we did not land https://github.com/zalando/skipper/pull/2080

Lets wait for it maybe?

File: dataclients/kubernetes/clusterstate.go [hunk 6]
```diff
@@ -55,6 +51,26 @@ func (state *clusterState) GetEndpointsByService(namespace, name, protocol strin
...
+	}
+
+	ep, ok := state.endpoints[epID.ResourceID]
+	if !ok {
+		return nil
+	}
+
+	targets := ep.targetsByServicePort(protocol, servicePort)
+	sort.Strings(targets)
+	state.cachedEndpoints[epID] = targets
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/clusterstate.go:65]: This does not look correct, either mutex should be used or cache should not be updated

File: dataclients/kubernetes/clusterstate.go (hunk 5 above)
Comment (szuecs) [reply dataclients/kubernetes/clusterstate.go:26]: ok

File: dataclients/kubernetes/clusterstate.go (hunk 6 above)
Comment (szuecs) [reply dataclients/kubernetes/clusterstate.go:71]: fixed

**Comment by szuecs** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:
